### PR TITLE
zfs-2.2.6 patchset

### PR DESCRIPTION
**Pull request by tonyhutter**
### Motivation and Context
Proposed patchset for zfs-2.2.6.

### Description
Support 6.10 kernel, misc bugfixes.

### How Has This Been Tested?
Buildbot will test

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [ ] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [ ] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Review by robn — APPROVED** [review]
This one is tight, I like it. Good stuff!

**Comment by tonyhutter** [comment]
I also need 767b37019 bdf4d6be1 - let me fix that.

**Comment by wmmur** [comment]
I must say that zfs-master has fewer bugs than zfs-2.2.5 at release! linux-6.11.0.rc3 and rc4 works compilation and compilation of zfs-dkms on debian testing and devuan ceres, I have been using since adding patch linux-6.11 by @robn. I tested this patch on zfs-2.2.5 build with no errors! compilation zfs-dkms error incompatible kernel linux 6.11.
0 rc3 and rc4. zfs-master fixed zfs-dkms compilation for kernel linux-6.10.6-rt. Can you add to patch zfs-2.2.6 the removal of "DECLARE_EVENT_CLASS", @tonyhutter you also had a problem with this garbage. 
also, I had a weird error with make native-deb-utils I had to use "chmod +x" for the "/etc/init.d/zfs*" files. reason: zfs-
import error access denied.

**Comment by wmmur** [comment]
devuan ceres zfs-2.2.6-hutter make native-deb-utils without errors. zfs-dkms compilation for linux-6.10.6-rt:

/var/lib/dkms/zfs/2.2.6/build/module/os/linux/zfs/zvol_os.c:1193:9: warning: data definition has no type or storage class
 1193 |         blk_queue_set_write_cache(queue, B_TRUE);
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~
/var/lib/dkms/zfs/2.2.6/build/module/os/linux/zfs/zvol_os.c:1193:9: error: type defaults to ‘int’ in declaration of ‘blk_queue_set_write_cache’ [-Werror=implicit-int]
/var/lib/dkms/zfs/2.2.6/build/module/os/linux/zfs/zvol_os.c:1193:9: warning: parameter names (without types) in function declaration
/var/lib/dkms/zfs/2.2.6/build/module/os/linux/zfs/zvol_os.c:1193:9: error: conflicting types for ‘blk_queue_set_write_cache’; have ‘int()’
In file included from /var/lib/dkms/zfs/2.2.6/build/include/os/linux/spl/sys/uio.h:31,
                 from /var/lib/dkms/zfs/2.2.6/build/include/os/linux/spl/sys/sunddi.h:28,
                 from /var/lib/dkms/zfs/2.2.6/build/include/sys/zfs_context.h:69,
                 from /var/lib/dkms/zfs/2.2.6/build/include/sys/dmu.h:46,
                 from /var/lib/dkms/zfs/2.2.6/build/include/sys/dataset_kstats.h:31,
                 from /var/lib/dkms/zfs/2.2.6/build/module/os/linux/zfs/zvol_os.c:25:
/var/lib/dkms/zfs/2.2.6/build/include/os/linux/kernel/linux/blkdev_compat.h:79:1: note: previous definition of ‘blk_queue_set_write_cache’ with type ‘void(struct request_queue *, bool)’ {aka ‘void(struct request_queue *, _Bool)’}
   79 | blk_queue_set_write_cache(struct request_queue *q, bool on)
      | ^~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /var/lib/dkms/zfs/2.2.6/build/include/os/linux/spl/sys/uio.h:30:
/usr/src/linux-headers-6.10.6-common-rt/include/linux/blkdev.h:531:33: error: expected declaration specifiers or ‘...’ before numeric constant
  531 | #define QUEUE_FLAG_IO_STAT      7       /* do disk/partitions IO accounting */
      |                                 ^
/var/lib/dkms/zfs/2.2.6/build/module/os/linux/zfs/zvol_os.c:1194:28: note: in expansion of macro ‘QUEUE_FLAG_IO_STAT’
 1194 |         blk_queue_flag_set(QUEUE_FLAG_IO_STAT, queue);
      |                            ^~~~~~~~~~~~~~~~~~
/var/lib/dkms/zfs/2.2.6/build/module/os/linux/zfs/zvol_os.c:1194:48: error: unknown type name ‘queue’
 1194 |         blk_queue_flag_set(QUEUE_FLAG_IO_STAT, queue);
      |                                                ^~~~~
/var/lib/dkms/zfs/2.2.6/build/module/os/linux/zfs/zvol_os.c:1196:1: error: expected identifier or ‘(’ before ‘}’ token
 1196 | }
      | ^
/var/lib/dkms/zfs/2.2.6/build/module/os/linux/zfs/zvol_os.c:1197:2: error: #endif without #if
 1197 | #endif
      |  ^~~~~
  AS [M]  /var/lib/dkms/zfs/2.2.6/build/module/icp/asm-x86_64/blake3/blake3_avx512.o
cc1: some warnings being treated as errors
make[5]: *** [/usr/src/linux-headers-6.10.6-common-rt/scripts/Makefile.build:249: /var/lib/dkms/zfs/2.2.6/build/module/os/linux/zfs/zvol_os.o] Błąd 1
make[5]: *** Oczekiwanie na niezakończone zadania....
make[4]: *** [/usr/src/linux-headers-6.10.6-common-rt/Makefile:1959: /var/lib/dkms/zfs/2.2.6/build/module] Błąd 2
make[3]: *** [/usr/src/linux-headers-6.10.6-common-rt/Makefile:252: __sub-make] Błąd 2
make[3]: Opuszczenie katalogu '/usr/src/linux-headers-6.10.6-rt-amd64'
make[2]: *** [Makefile:56: modules-Linux] Błąd 2
make[2]: Opuszczenie katalogu '/var/lib/dkms/zfs/2.2.6/build/module'
make[1]: *** [Makefile:973: all-recursive] Błąd 1
make[1]: Opuszczenie katalogu '/var/lib/dkms/zfs/2.2.6/build'
make: *** [Makefile:877: all] Błąd 2

**Comment by wmmur** [comment]
devuan ceres kernel linux-6.11.0 rc4 zfs-dkms build error
/var/lib/dkms/zfs/2.2.6/build/module/os/linux/zfs/zvol_os.c:1196:1: error: expected identifier or ‘(’ before ‘}’ token
 1196 | }
      | ^
/var/lib/dkms/zfs/2.2.6/build/module/os/linux/zfs/zvol_os.c:1197:2: error: #endif without #if
 1197 | #endif
      |  ^~~~~
  AS [M]  /var/lib/dkms/zfs/2.2.6/build/module/icp/asm-x86_64/blake3/blake3_sse41.o
make[5]: *** [scripts/Makefile.build:244: /var/lib/dkms/zfs/2.2.6/build/module/os/linux/zfs/zvol_os.o] Błąd 1
make[5]: *** Oczekiwanie na niezakończone zadania....
make[4]: *** [/usr/src/linux-headers-6.11.0-061100rc4-generic/Makefile:1931: /var/lib/dkms/zfs/2.2.6/build/module] Błąd 2
make[3]: *** [Makefile:224: __sub-make] Błąd 2
make[3]: Opuszczenie katalogu '/usr/src/linux-headers-6.11.0-061100rc4-generic'
make[2]: *** [Makefile:56: modules-Linux] Błąd 2
make[2]: Opuszczenie katalogu '/var/lib/dkms/zfs/2.2.6/build/module'
make[1]: *** [Makefile:973: all-recursive] Błąd 1
make[1]: Opuszczenie katalogu '/var/lib/dkms/zfs/2.2.6/build'
make: *** [Makefile:877: all] Błąd 2

**Comment by wmmur** [comment]
in the file "module/os/linux/zfs/zvol_os.c" on line 1184 the following is missing:
#ifndef HAVE_BLK_ALLOC_DISK_2ARG

**Comment by jumbi77** [comment]
Maybe too early (?) but maybe include this recent fix: https://github.com/openzfs/zfs/commit/2420ee6e12cb4bc4918fc88d44d59b486b86e58b

**Comment by Harry-Chen** [comment]
I looked back at #16359 and noticed some suggestions that were not in 2.2.5: #16376, #16409, #16410, #16411.

Also there are two of my early fixes that could be quickly pulled: #15557, #15623.

**Comment by jumbi77** [comment]
> I looked back at #16359 and noticed some suggestions that were not in 2.2.5: #16376, #16409, #16410, #16411.
> 
> Also there are two of my early fixes that could be quickly pulled: #15557, #15623.

Beside that, I just got a general question: I tried using GitHub compare [master->ifs-2.2.6-staging](https://github.com/openzfs/zfs/compare/zfs-2.2.6-staging...master). It shows a lot of commits, but I know at least some of them are included in 2.2.x, just as another commit-hash I think... is there a better/way way to really check which underlying changes/commits are not included from master?

**Comment by robn** [comment]
@jumbi77 `git cherry zfs-2.2.6-staging master -v` is good for this (show commits that aren't in `zfs-2.2.6-staging` but not in `master`, by content).

**Comment by wmmur** [comment]
@robn, @behlendorf  what changes were made to zfs-master when adding patch linux-6.11 that zfs-master compilation of zfs-dkms on linux-6.10.6-rt works but compilation of zfs-dkms on zfs-2.2.6 does not work?

File: module/os/linux/zfs/zvol_os.c [hunk 1]
```diff
@@ -1169,6 +1185,14 @@ zvol_queue_limits_apply(zvol_queue_limits_t *limits,
 	blk_queue_max_segments(queue, limits->zql_max_segments);
```
Comment (wmmur) [inline module/os/linux/zfs/zvol_os.c:1185]: missing #ifndef HAVE_BLK_ALLOC_DISK_2ARG

File: module/os/linux/zfs/zvol_os.c
Comment (wmmur) [inline module/os/linux/zfs/zvol_os.c:1]: miss #ifndef HAVE_BLK_ALLOC_DISK_2ARG

File: module/os/linux/zfs/zvol_os.c (hunk 1 above)
Comment (wmmur) [inline module/os/linux/zfs/zvol_os.c:1185]: miss #ifndef HAVE_BLK_ALLOC_DISK_2ARG

**Comment by jaen** [comment]
Any chance to get https://github.com/openzfs/zfs/pull/16402 in for this, or building own modules still the only choice?

**Comment by rincebrain** [comment]
> > I looked back at #16359 and noticed some suggestions that were not in 2.2.5: #16376, #16409, #16410, #16411.
> > Also there are two of my early fixes that could be quickly pulled: #15557, #15623.
> 
> Beside that, I just got a general question: I tried using GitHub compare [master->ifs-2.2.6-staging](https://github.com/openzfs/zfs/compare/zfs-2.2.6-staging...master). It shows a lot of commits, but I know at least some of them are included in 2.2.x, just as another commit-hash I think... is there a better/nice way to really check which underlying changes/commits are not included from master?

github lets you select a branch, you could diff against zfs-2.2.5 or zfs-2.2-release.

**Comment by tonyhutter** [comment]
@jumbi77 taskq stats aren't in 2.2.x, so I skipped 2420ee6
@Harry-Chen I pulled all those in except https://github.com/openzfs/zfs/pull/16411.
@jaen I pulled in #16402

**Pull request merged**